### PR TITLE
#19 included new InteractiveParameterResolver, added a new readLine …

### PR DESCRIPTION
**Pull request by camilojc**
…method to the Shell interface and included an interactive and masked boolean parameters to the ShellOption annotation

Also includes:
changed the ExitRequest for an ExitResolver and added an ExitResultHandler to avoid using exception handling in the exit flow
removed the ApplicationContext parameter from the MethodTargetResolver (should now be included via injection)
made the StandardParameterResolver an @Component and removed its definition from the Bootstrap
updated JLine to version 3.2.0
added .gitattributes to avoid eol conversion in Windows machines for reference files used in tests

### CODE REVIEW

File: src/main/java/org/springframework/shell2/standard/ShellOption.java [hunk 1]
```diff
@@ -27,19 +27,20 @@
  *
  * @author Eric Bottard
  * @author Florent Biville
+ * @author Camilo Gonzalez
  */
 @Documented
 @Retention(RetentionPolicy.RUNTIME)
-@Target(ElementType.PARAMETER)
+@Target({ ElementType.PARAMETER, ElementType.FIELD })
```
Comment (camilojc) [inline src/main/java/org/springframework/shell2/standard/ShellOption.java]: added in preparation to allow for @ShellOption inside "argument" classes similar to JCommander, e.g.:

```java
@ShellMethod(...)
public void doSomething(ArgumentHolder args) {...}

private static class ArgumentHolder {
   @ShellOption(...)
   private String argument1;
}
```

File: src/main/java/org/springframework/shell2/standard/InteractiveParameterResolver.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+			return conversionService.convert(readValue, TypeDescriptor.valueOf(String.class),
+					new TypeDescriptor(methodParameter));
+		}
+	}
+
+	private boolean isParameterMasked(MethodParameter parameter) {
+		return parameter.getParameterAnnotation(ShellOption.class).masked();
+	}
+	
+	private Stream<String> getKeysForParameter(Method method, int index) {
```
Comment (camilojc) [inline src/main/java/org/springframework/shell2/standard/InteractiveParameterResolver.java]: brought this from the StandardParameterResolver initially, but need to find a common place to put it - looks like a ShellOptionUtility or ShellOptionHelper class (?)

File: .gitattributes [hunk 3]
```diff
@@ -0,0 +1 @@
+src/test/resources/org/springframework/shell2/commands/HelpTest*.txt eol=lf
```
Comment (ericbottard) [inline .gitattributes]: We could apply eol=lf to the whole project, as a separate PR

File: pom.xml [hunk 4]
```diff
@@ -30,14 +30,14 @@
 			<artifactId>spring-boot-starter-test</artifactId>
 		</dependency>
 		<dependency>
-			<groupId>jline</groupId>
+			<groupId>org.jline</groupId>
```
Comment (ericbottard) [inline pom.xml]: This upgrade is valuable in itself and can be the subject of a dedicated PR

File: src/main/java/org/springframework/shell2/MethodTargetResolver.java [hunk 5]
```diff
@@ -18,13 +18,16 @@
...
+ * Defines a {@link MethodTargetResolver} that will return a mapping from Command -> {@link MethodTarget}.
+ * </p>
+ * 
+ * @author Eric Bottard
+ * @author Camilo Gonzalez
  */
 public interface MethodTargetResolver {
 
-	public Map<String, MethodTarget> resolve(ApplicationContext context);
+	public Map<String, MethodTarget> resolve();
```
Comment (ericbottard) [inline src/main/java/org/springframework/shell2/MethodTargetResolver.java]: Could you make this removal a dedicated PR?

**Comment by camilojc** [comment]
@ericbottard thanks for the comments. After our chat I've decided to close this PR and open individual ones for the enhancements I've been adding. I'll be resubmitting them in the next couple of days, ideally on top of the project subdivision into submodules (to avoid merge conflicts :) )

File: src/main/java/org/springframework/shell2/MethodTargetResolver.java (hunk 5 above)
Comment (camilojc) [inline src/main/java/org/springframework/shell2/MethodTargetResolver.java]: https://github.com/ericbottard/spring-shell2/issues/39 to keep track of it

File: pom.xml (hunk 4 above)
Comment (camilojc) [inline pom.xml]: I think you've already applied this change to master 👍
